pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1181,36 +1181,41 @@ func (c *PodAffinityChecker) InterPodAffinityMatches(pod *v1.Pod, meta algorithm
 	return true, nil, nil
 }
 
-// podMatchesPodAffinityTerms checks if the "targetPod" matches the given "terms"
-// of the "pod" on the given "nodeInfo".Node(). It returns three values: 1) whether
-// targetPod matches all the terms and their topologies, 2) whether targetPod
-// matches all the terms label selector and namespaces (AKA term properties),
-// 3) any error.
-func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulercache.NodeInfo, terms []v1.PodAffinityTerm) (bool, bool, error) {
-	if len(terms) == 0 {
-		return false, false, fmt.Errorf("terms array is empty")
-	}
-	props, err := getAffinityTermProperties(pod, terms)
+// podMatchesPodAffinityTerms checks if the "targetPod" matches the "checkAffinity (true:affinity,false:anti-affinity)"
+// of the "pod" on the given "nodeInfo".Node(). It returns four values: 1) whether
+// targetPod matches any of the affnityTerms and their topologies, 2) whethertargetPod matches
+// all the affnityTerms and their topologies, 3) whether targetPodmatches all the terms label
+// selector and namespaces (AKA term properties), 4) any error.
+func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulercache.NodeInfo, checkAffinity bool) (bool, bool, bool, error) {
```

Comment:
Why do we need to return so many bools? We know whether we are checking affinity or anti-affinity. So, this function can return whether the pods match or not and an error. So, the return value should be just `(bool, error)`.

```suggestion
func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod, nodeInfo *schedulercache.NodeInfo, checkAffinity bool) (bool, error) {
	matchedTerms, affinityTermsNum, err := getMatchingAffinityTerms(pod, targetPod, checkAffinity)
	if err != nil {
		return false, err
	}

	if checkAffinity && len(matchedTerms) != affinityTermsNum {
		return false, err
	}
		
	targetPodNode, err := c.info.GetNodeInfo(targetPod.Spec.NodeName)
	if err != nil {
		return false, false, err
	}

       // Check topology of the terms.
	for _, term := range matchedTerms {
		if len(term.TopologyKey) == 0 {
			return false, fmt.Errorf("empty topologyKey is not allowed except for PreferredDuringScheduling pod anti-affinity")
		}
		topologyMatched := priorityutil.NodesHaveSameTopologyKey(nodeInfo.Node(), targetPodNode, term.TopologyKey)
		// if any rules matches for anti-affinity, it is a violation.
		if topologyMatched && !checkAffinity {
			return false, nil
		}
		// if any rules does not match for affinity, it is a non-match.
		if !topologyMatch && checkAffinity {
			return false, nil
		}
	}
	return true, nil
}
```